File: frontend/src/services/api/grammar.ts
import { apiClient } from "@/services/api/client";
import type { GrammarTopicDetail, GrammarTopicListItem } from "@/types/grammar";

export interface GrammarFilters {
  category?: string;
  search?: string;
}

export const grammarApi = {
  list: async (filters: GrammarFilters = {}): Promise<GrammarTopicListItem[]> => {
    const { data } = await apiClient.get<GrammarTopicListItem[]>("/grammar", { params: filters });
    return data;
  },

  categories: async (): Promise<string[]> => {
    const { data } = await apiClient.get<string[]>("/grammar/categories");
    return data;
  },

  getBySlug: async (slug: string): Promise<GrammarTopicDetail> => {
    const { data } = await apiClient.get<GrammarTopicDetail>(`/grammar/${slug}`);
    return data;
  },
};

File: backend/app/api/v1/grammar.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.repositories.grammar_repository import GrammarRepository
from app.schemas.grammar import GrammarTopicDetail, GrammarTopicListItem
from app.services.grammar_service import GrammarService

router = APIRouter(prefix="/grammar", tags=["grammar"])


def get_service(db: Session = Depends(get_db)) -> GrammarService:
    return GrammarService(GrammarRepository(db))


@router.get("", response_model=list[GrammarTopicListItem])
def list_topics(
    category: str | None = None,
    search: str | None = None,
    service: GrammarService = Depends(get_service),
):
    return service.list(category=category, search=search)


@router.get("/categories", response_model=list[str])
def list_categories(service: GrammarService = Depends(get_service)):
    return service.list_categories()


@router.get("/{slug}", response_model=GrammarTopicDetail)
def get_topic(slug: str, service: GrammarService = Depends(get_service)):
    return service.get_by_slug_or_404(slug)
